### ISSUE TITLE
[SOUND] Wouln't be better to have separated layers for diferent sound tracks?

### ISSUE THREAD

**Issue by Kaikogf** [pencil2d/pencil#749]
### --Issue Summary--
I should have start this comment by asking if it is normal to have several sound tracks imported to the same layer?
If it is Ok, then I will close/delete this post 👍

**Comment by CandyFace** [comment]
Why would you want a layer per sound? that would kind of defeat the purpose of having frames for sound effects. it would also make quite a messy timeline when you're working with lots of sounds.

Multiple sound effects is perfectly fine on a single layer.

**Comment by Jose-Moreno** [comment]
@CandyFace actually this is fairly common in video editing. Having multiple layers with different sounds help you "layer" the sounds when creating complex ambience motifs. Although it's better to think about it as having tracks. Sometimes the editor use a single track for a single sound, sometimes it uses them for sound categories like "soundtrack" or "foley" or "dialogue".

For an animation package though I don't see the harm in allowing people to have multiple sound layers either. In TVPaint you can have several sound layers (separated from your drawing layers though) to have dialogue and soundfx separated.

We have a working sound layer. If the user wants to have 20 different sound layers with individual sounds or just one layer it's several sounds, they already can do that and it's their choice.

The latter option is a bit tricky in Pencil2D though because we don't have the ability to trim sounds to accommodate for specific lengths of the sound saple. You see it a lot in animation voice recording. The whole dialogue from a single session is recorded (it's cheaper) and then edited to get individual performances from the actors.

Since the sounds in Pencil2D are always embedded inside a single frame container it's hard to use unedited samples, or stock sounds without editing them outside in something like Audacity, but sometimes you just want to get a specific beat of a sound, and you need the visual cue to get it right, so hopping back and forth from your audio editor is very time consuming.

So in this case fitting multiple sounds inside a single layer is actually harder than having multiple layers with a single sound in Pencil2D.

**Comment by Kaikogf** [comment]
When I first post that I was meanly thinking about the poor colour contrast of the sound tracks showing on the timeline. My eyes were not adapted to that but now I understand what was my idea: I wasn't able to see where each track started. Now I realized that  **_IF_** I select the track that is most to the right I can see all the other.

The more sound tracks you have the more difficult it get to clearly identify where is which, because of the dimming of the sound layers

There is no reference or anything to allow us to tell "here start this track"... "there the other one". Selecting the most to the left will "hide" all the others. What about having 10 sound tracks?


Well, this is just an idea but as an enhancement the developers could certainly add a small scrub to the beginning of each track. That would surely help when there are multiple sound tracks 👍

**Comment by Kaikogf** [comment]
@Jose-Moreno 

I just read what you posted. I agree

**Comment by CandyFace** [comment]
@Jose-Moreno 
> actually this is fairly common in video editing. Having multiple layers with different sounds help you "layer" the sounds when creating complex ambience motifs. Although it's better to think about it as having tracks. Sometimes the editor use a single track for a single sound, sometimes it uses them for sound categories like "soundtrack" or "foley" or "dialogue".

That's not what I meant, I know that having individual tracks for, music, foley, sound effects and dialogue is common for better overview but that's already possible. From the programs i've used eg, Audacity and Ardour, it's still possible to have multiple sounds on a track though, so If people want to have multiple sounds on a layer in Pencil2D, then I don't see why we should remove that freedom.

> Since the sounds in Pencil2D are always embedded inside a single frame container it's hard to use unedited samples, or stock sounds without editing them outside in something like Audacity, but sometimes you just want to get a specific beat of a sound, and you need the visual cue to get it right, so hopping back and forth from your audio editor is very time consuming.

We could try to implement such that the timeline will check if imported sound files has been modified outside pencil, then compare against a unique hash id or something and reimport if it doesn't match. That should make sound editing easier until we rework the timeline and allow direct sample manipulation.

@Kaikogf 
> There is no reference or anything to allow us to tell "here start this track"... "there the other one". Selecting the most to the left will "hide" all the others. What about having 10 sound tracks?

This could be fixed by always showing tracks on top of each other, so even if you selected a specific track, other tracks which would be overlapping would still be shown.

A marker for tracks and possibly frames is an interesting idea though, I'll see if I can incorporate that into my timeline concept in #533

**Comment by Jose-Moreno** [comment]
@Candyface Thanks for providing insight as always! Actually by reading your response it becomes clear that what we actually need is to rewrite the timeline. Just thinking logically, if the timeline rewrite will help structure these issues, then let's  wait for that. There's no need to patch things right now if those changes would become nullified by rewriting the timeline. If the changes will persist then let's plan out these details properly.

To me with your timeline proposal, the trick for sound lies with the exposure system. The layer frame "container " will adapt it's exposure to match the sound length, and the user will be able to increase or decrease the exposure of the container in frame units to shorten the sound or lengthen the container (thus having empty frames) or even inserting new frames with new exposure to the same effect.

Having said this It becomes very apparent that we really need to the timeline to be rewritten at least before 2018. I'm going to look soon into the Software Freedom Conservancy group https://sfconservancy.org/ maybe they can help us administering donations and stuff so people will be really willing to back this project knowign their money will be on the right hands, aside form the bountysource / liberapay methods.

They are helping out other high profile software, so perhaps this will be able to assist pencil2D into finally evolving. I'll read about their requirements and submit an information inquiry hopefully this weekend, or when I have some free time. Let's see what happens.

**Comment by CandyFace** [comment]
@Jose-Moreno 
No problem :) 
Indeed we need to rewrite the timeline, that's pretty much what i'm waiting for too but
setting a deadline on something like this is difficult, especially if you want someone from outside that doesn't know the coding structure and there's hardly any documentation the source code. @scribblemaniac has said that he wants to rewrite it himself, so we'll just have to wait for that to happen, whether that's going to be done before we hit 2018, is 100% up to him.

**Comment by Jose-Moreno** [comment]
@CandyFace Of course, didn't meant it to rush the task. I really want you guys to get some kind of compensation though. You've given a lot to the software so I think it's just measure. Hopefully we can count on next year being great for Pencil2D :)

**Comment by scribblemaniac** [comment]
I will do my best to get the timeline rewrite done, but it's difficult to say exactly how long this will take. If things continue at the current rate, I'm positive I can get it done for the v0.7 release...which would be in 2021 😝. Jokes aside, I will aim to finish a significant portion of the new timeline by the end of the year. This issue and other similar ones do demonstrate the need for a timeline rewrite, but they also demonstrate the large number of design considerations that should be taken into account in the early stages and the amount of features that will need to be made in development stages. To clarify slightly, I do not aim to achieve this all on my own, I just wish to get the foundation of it done in a way that can hopefully better support things like this. At the very least I should be accepting PRs well before December. I will try to go over everything brought up so far and post an update to #533.

**Comment by chchwy** [comment]
@scribblemaniac  I'm thinking to add a timeline-rewrite branch in the main repository so everyone who wants to contribute could join you, any thoughts?

**Comment by scribblemaniac** [comment]
@chchwy As noted in [this comment](https://github.com/pencil2d/pencil/issues/533#issuecomment-305409281), I would like to get a bit more done before having other people contribute to it. This gives me the chance to make large changes if necessary without worrying about breaking things others are working on. However once that's cleared, as long as I have write access I think that would be a wonderful idea ^_^ I will keep you posted.

**Comment by chchwy** [comment]
@scribblemaniac good to know  :)
